src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.cs

```diff
@@ -12,37 +14,194 @@ internal static partial class ClrThreadPool
-        // TODO: Compare with CoreCLR implementation and ensure this has the same guarantees.
         public static bool SetMinThreads(int threads)
         {
+            s_maxMinThreadLock.Acquire();
+            bool success;
             if (threads < 0 || threads > s_maxThreads)
             {
-                return false;
+                success = false;
             }
             else
             {
                 s_minThreads = threads;
-                return true;
+
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_counts);
+                while(counts.maxWorking < s_minThreads)
+                {
+                    ThreadCounts newCounts = counts;
+                    newCounts.maxWorking = (short)s_minThreads;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_counts, newCounts, counts);
+                    if (oldCounts == counts)
+                    {
+                        counts = newCounts;
+
+                        if(newCounts.maxWorking > oldCounts.maxWorking && ThreadPool.GetQueuedWorkItems().Any())
+                        {
+                            WorkerThread.MaybeAddWorkingWorker();
+                        }
+                    }
+                    else
+                    {
+                        counts = oldCounts;
+                    }
+                }
+                success = true;
             }
+            s_maxMinThreadLock.Release();
+            return success;
         }
 
         public static int GetMinThreads() => s_minThreads;
 
         public static bool SetMaxThreads(int threads)
         {
+            s_maxMinThreadLock.Acquire();
+            bool success;
             if (threads < s_minThreads || threads == 0)
             {
-                return false;
+                success = false;
             }
             else
             {
                 s_maxThreads = threads;
-                return true;
+
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_counts);
+                while (counts.maxWorking > s_maxThreads)
+                {
+                    ThreadCounts newCounts = counts;
+                    newCounts.maxWorking = (short)s_maxThreads;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_counts, newCounts, counts);
+                    if (oldCounts == counts)
+                    {
+                        counts = newCounts;
+                    }
+                    else
+                    {
+                        counts = oldCounts;
+                    }
+                }
+                success = true;
             }
+            s_maxMinThreadLock.Release();
+            return success;
         }
 
         public static int GetMaxThreads() => s_maxThreads;
+
+        internal static bool NotifyWorkItemComplete()
+        {
+            Interlocked.Increment(ref s_completionCount);
```

Comment:
This is a hot path (called once per work item) and the original code is using a thread-local value for what looks like performance purposes. Could you please add a TODO here to check the perf? I suspect the interlocked increment is fine.